### PR TITLE
InmutableString working

### PR DESCRIPTION
**Pull request by ZaruWright**
A first version of InmutableString class. Also I have created an another class that implements a log to debug.

### CODE REVIEW

File: doc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+# check if Doxygen is installed
+find_package(Doxygen)
+if (DOXYGEN_FOUND)
+    # set input and output files
+    set(DOXYGEN_IN ${CMAKE_SOURCE_DIR}/doc/Doxyfile.in)
+    set(DOXYGEN_OUT ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile)
+
+    # request to configure the file
+    configure_file(${DOXYGEN_IN} ${DOXYGEN_OUT} @ONLY)
+    message("Doxygen build started")
```
Comment (Manu343726) [inline doc/CMakeLists.txt]: remove

File: doc/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    # note the option ALL which allows to build the docs together with the application
+    add_custom_target( doc ALL
+        COMMAND ${DOXYGEN_EXECUTABLE} ${DOXYGEN_OUT}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+        COMMENT "Generating API documentation with Doxygen"
+        VERBATIM )
+else (DOXYGEN_FOUND)
+  message("Doxygen need to be installed to generate the doxygen documentation")
```
Comment (Manu343726) [inline doc/CMakeLists.txt]: `message(FATAL_ERROR "foo")` aborts cmake processing with an error message

File: doc/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+endif (DOXYGEN_FOUND)
+
+# if(BUILD_DOCUMENTATION)
```
Comment (Manu343726) [inline doc/CMakeLists.txt]: Remove commented out code

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,9 @@
 project(siminusminus)
 cmake_minimum_required(VERSION 2.8)
 
+# Debug macros
+set(CMAKE_CXX_FLAGS_DEBUG -D"_DEBUG")
```
Comment (Manu343726) [inline CMakeLists.txt]: Not needed, since CMake already sets the "standard" `NDEBUG` macro on Release builds. The proper "standard" way to check if you're building on release/debug is to check if `NDEBUG` is defined:

``` cpp
#ifndef NDEBUG
    // Debug
#else
    // Release
#endif // NDEBUG
```

See http://en.cppreference.com/w/c/error/assert

Also, is better to set flags and other properties per target, not globally.

File: include/siminusminus/containers/inmutablestring.hpp [hunk 5]
```diff
@@ -1,6 +1,9 @@
 #ifndef SIMINUSMINUS_CONTAINERS_INMUTABLESTRING_HPP
 #define SIMINUSMINUS_CONTAINERS_INMUTABLESTRING_HPP
 
+#include "siminusminus/utils/debugutilities.hpp"
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: Use angle brackets (`<>`) for global includes, `""` for local includes. Since you're including the header fully qualifying it (Which is the right way in this situation IMO since the header will be exposed to the public), use angle brackets.

File: include/siminusminus/containers/inmutablestring.hpp [hunk 6]
```diff
@@ -28,11 +31,115 @@ namespace containers
...
+
+	char* _string;  // String value
+	size_t _length; // length of the string
 };
 
-}
+} // NAMESPACE containers
 
-}
+} // NAMESPACE cmm
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: A good practice I've to do more, but since the `namespace` keyword is lowercase, let's write it lowercase ;)

File: include/siminusminus/utils/debugutilities.hpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+{
+
+/**
+ * \ingroup utils
+ * \brief Debug helping class.
+ *
+ * Help class to debug our libraries. We set with CMAKE the "_DEBUG" macro only when
+ * the CMAKE_BUILD_TYPE is "Debug".
+ */
+class DebugUtilities
```
Comment (Manu343726) [inline include/siminusminus/utils/debugutilities.hpp]: Great idea, but I would go with a proper logging framework instead. Let's discuss in the next class

File: mbuild.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+build=$1
+
+DIRECTORY=./_build
+if [ ! -d "$DIRECTORY" ]; then
+	mkdir "$DIRECTORY"
+fi
+
+cd "$DIRECTORY"
+if [ "$build" == "release" ]; then
```
Comment (Manu343726) [inline mbuild.sh]: I think you would like @floooh's [fips](https://github.com/floooh/fips)

File: src/utils/debugutilities.cpp [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#include <siminusminus/utils/debugutilities.hpp>
+
+namespace cmm{
+namespace utils{
+
+	void cmm::utils::DebugUtilities::log(const std::string& string, bool newline)
+	{
+		#ifdef _DEBUG
+			std::cout << string << std::endl;
+			if (newline) // We do this to not mix '/n' and std::endl
```
Comment (Manu343726) [inline src/utils/debugutilities.cpp]: ???? Using `\n` is fine

File: test/containers/inmutablestring_test.cpp [hunk 10]
```diff
@@ -4,4 +4,112 @@
...
 
-// TODO: InmutableString unit tests here
+/////////////////
+// Initialization
+/////////////////
+
+TEST(initialization, voidConstructor)
+{
+	InmutableString str1;
+	EXPECT_TRUE(str1 == "");
```
Comment (Manu343726) [inline test/containers/inmutablestring_test.cpp]: Use `EXPECT_EQ(expected, given)`, i.e.:

``` cpp
EXPECT_EQ("", str1);
```

File: test/containers/inmutablestring_test.cpp [hunk 11]
```diff
@@ -4,4 +4,112 @@
...
+}
+
+TEST(operators, getNElement)
+{
+	InmutableString str("Hello");
+	EXPECT_TRUE(str[0] == 'H' &&
+		        str[1] == 'e' &&
+		        str[2] == 'l' &&
+		        str[3] == 'l' &&
+		        str[4] == 'o');
```
Comment (Manu343726) [inline test/containers/inmutablestring_test.cpp]: Don't do this (In addition to the comparison thing). The point of using the macros is to have proper diagnostics when the test fails. Doing it like this you will never know why your test failed since you will not know the value of each tested string element.

File: test/containers/inmutablestring_test.cpp [hunk 12]
```diff
@@ -4,4 +4,112 @@
...
+	InmutableString str1("HelloWorld");
+	InmutableString str2("HelloWorld");
+	EXPECT_TRUE(str1 == str2);
+}
+
+TEST(operators, greatherThan)
+{
+	InmutableString str1("Hello World");
+	InmutableString str2("HelloWorld");
+	EXPECT_TRUE(str1 > str2);
```
Comment (Manu343726) [inline test/containers/inmutablestring_test.cpp]: `EXPECT_GT(str1, str2)`. Anyway, what's this? Lexicographical comparison?

File: test/containers/inmutablestring_test.cpp [hunk 13]
```diff
@@ -4,4 +4,112 @@
...
+}
+
+TEST(operators, lessThan)
+{
+	InmutableString str1("HelloWorld");
+	InmutableString str2("Hello World");
+	EXPECT_TRUE(str1 < str2);
```
Comment (Manu343726) [inline test/containers/inmutablestring_test.cpp]: `EXPECT_LT()`

File: test/containers/main.cpp [hunk 14]
```diff
@@ -1,9 +1,15 @@
 #include <gmock/gmock.h>
+#include <siminusminus/containers/inmutablestring.hpp>
 
 using namespace ::testing;
+using namespace cmm::containers;
```
Comment (Manu343726) [inline test/containers/main.cpp]: Remove

File: test/containers/main.cpp [hunk 15]
```diff
@@ -1,9 +1,15 @@
 #include <gmock/gmock.h>
+#include <siminusminus/containers/inmutablestring.hpp>
```
Comment (Manu343726) [inline test/containers/main.cpp]: Remove

File: test/containers/main.cpp [hunk 16]
```diff
@@ -1,9 +1,15 @@
...
 
 using namespace ::testing;
+using namespace cmm::containers;
 
 int main(int argc, char **argv)
 {
-    ::testing::InitGoogleTest(&argc, argv);
+	// #ifdef _DEBUG
```
Comment (Manu343726) [inline test/containers/main.cpp]: Remove. NEVER NEVER NEVER commit commented code

File: test/containers/main.cpp [hunk 17]
```diff
@@ -1,9 +1,15 @@
...
+	// 	std::cout << "Hola" << std::endl;
+	::testing::InitGoogleTest(&argc, argv);
     return RUN_ALL_TESTS();
+    
+	// return 0;
```
Comment (Manu343726) [inline test/containers/main.cpp]: Remove

File: src/containers/inmutablestring.cpp [hunk 18]
```diff
@@ -1,5 +1,143 @@
...
+	_length = newLength;
+
+	std::strcpy(_string, newString); // Copy the string
+}
+
+void cmm::containers::InmutableString::stringLog() const
+{
+	cmm::utils::DebugUtilities::log(std::string("String: " + this->toString()));
+	cmm::utils::DebugUtilities::log(std::string("String length: ") + std::to_string(_length), true);
+}
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: Note the casts are not needed since the other operands already are `std::string`s (There are `std::string` `operator+()` overloads for `std::string` being the left or right operand). Also, since you're in a cpp file it's safe to do `using namespace cmm::utils;` and do `DebugUtilities::log()` instead

File: src/containers/inmutablestring.cpp [hunk 19]
```diff
@@ -1,5 +1,143 @@
...
+std::string cmm::containers::InmutableString::toString() const
+{
+	return std::string(_string);
+}
+
+void cmm::containers::InmutableString::createString(const size_t& newLength, const char* newString)
+{
+	size_t newSize = sizeof(char) * (newLength+1); // Inlude '/0'
+
+	_string = new char(newSize); // Get allocation memory for our string
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: I would check if `_string` is null first, this is a potential leak bug. If it is a precondition of the function, document it (And assert if it's met if you like).

Also, very very very very important: **Stop using `new` and `delete` this way, they don't do what you think**

If you read [the](http://en.cppreference.com/w/cpp/language/new) [docs](http://en.cppreference.com/w/cpp/memory/new/operator_new) carefully, you will note you're using the `operator new()` overload designed to allocate storage **for one object only**. It's fine in this case since you're tracking the size of the buffer, but I'm not sure you noticed the difference. Also, as a general rule, never call new/delete operators directly, use new/delete expressions instead. It worked because an string is a buffer of chars, but you would had been doomed if the exercise was to implement a generic array template (We will talk about this in depth in the next class).

File: include/siminusminus/containers/inmutablestring.hpp [hunk 20]
```diff
@@ -28,11 +31,115 @@ namespace containers
...
+     * Contructor from a C string.
+     * @param string: chain to create a InmutableString object.
+     */
+    InmutableString(const char* string);
+
+    /**
+     * Copy constructor. creates a new InmutableString object from another (istring).
+     * @param istring: The object which copy the data to create another InmutableSring object.
+     */
+    InmutableString(const InmutableString& istring);
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp:50]: Just curious: Why `istring`?

File: include/siminusminus/containers/inmutablestring.hpp (hunk 5 above)
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: Angle brackets

File: include/siminusminus/containers/inmutablestring.hpp [hunk 21]
```diff
@@ -28,11 +31,115 @@ namespace containers
...
+     * @param rhs: the right hand side of the operation.
+     */
+    bool operator>=(const InmutableString& rhs) const;
+
+    /**
+     * Returns true if the left hand side has less or equal length than rhs.
+     * @param rhs: the right hand side of the operation.
+     */
+    bool operator<=(const InmutableString& rhs) const;
+
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: Regarding the comparison operators, your semantics are confusing. No one would expect `str > str2` to check string lengths. Instead, the usual semantics of string comparison is lexicographical comparison.

 > **Hint:** if you want to fix this by implementing string lexicographical comparison, there's a [`std::lexicographical_compare()`](http://en.cppreference.com/w/cpp/algorithm/lexicographical_compare) algorithm in the standard library.

File: include/siminusminus/containers/inmutablestring.hpp [hunk 22]
```diff
@@ -28,11 +31,115 @@ namespace containers
...
+    /**
+     * Assignment operator. Copy the InmutableString of right hand side to 
+     * the left hand side.
+     * @param rhs: the right hand side of the operation.
+     */
+    InmutableString& operator=(const InmutableString& rhs);
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp:117]: No move constructor nor move assignment operator?

File: src/containers/inmutablestring.cpp [hunk 23]
```diff
@@ -1,5 +1,143 @@
...
+	cmm::utils::DebugUtilities::log(std::string("*** InmutableString::operator=(const InmutableString& rhs)."));
+
+	if (this == &rhs) // Points to the same data
+		return *this; // Not doing the assignment
+
+	if (this->_string) 
+		::operator delete(this->_string); // Erase data
+
+	this->_string = nullptr; // To take precautions
+	this->_length = 0;       // To take precautions
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: I would set these two values in a meaningful `setDefault()` function or something similar, and invoke it here

File: src/containers/inmutablestring.cpp [hunk 24]
```diff
@@ -1,5 +1,143 @@
...
+
+InmutableString& cmm::containers::InmutableString::operator=(const InmutableString& rhs)
+{
+	cmm::utils::DebugUtilities::log(std::string("*** InmutableString::operator=(const InmutableString& rhs)."));
+
+	if (this == &rhs) // Points to the same data
+		return *this; // Not doing the assignment
+
+	if (this->_string) 
+		::operator delete(this->_string); // Erase data
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: `delete _string` (See comments in `createString()`

File: src/containers/inmutablestring.cpp [hunk 25]
```diff
@@ -1,5 +1,143 @@
...
+
+bool cmm::containers::InmutableString::operator>=(const InmutableString& rhs) const
+{
+	return operator>(rhs) && operator==(rhs);
+}
+
+bool cmm::containers::InmutableString::operator<=(const InmutableString& rhs) const
+{
+	return operator<(rhs) && operator==(rhs);
+}
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: WTF I have never seen operators invoked in that obscure way

File: src/containers/inmutablestring.cpp [hunk 26]
```diff
@@ -1,5 +1,143 @@
...
+{
+	if (_string != nullptr)
+	{
+		cmm::utils::DebugUtilities::log(std::string("--- ~InmutableString()."));
+		stringLog();
+
+		delete(_string);
+		_string = nullptr;
+		_length = 0;
+	}
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: `setDefault()` thingy, as commented in the assignment operator

File: src/containers/inmutablestring.cpp [hunk 27]
```diff
@@ -1,5 +1,143 @@
...
+namespace containers{
 
-// TODO: InmutableString definition here
+////////////////
+// Constructors
+////////////////
+
+cmm::containers::InmutableString::InmutableString(): _string(nullptr), _length(0)
+{
+	createString(_length, "");
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: Doing a new of 0 element is even valid?

File: include/siminusminus/containers/inmutablestring.hpp [hunk 28]
```diff
@@ -28,11 +31,115 @@ namespace containers
 class InmutableString
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp:30]: Missing docs

File: include/siminusminus/containers/inmutablestring.hpp [hunk 29]
```diff
@@ -28,11 +31,115 @@ namespace containers
...
+
+private:
+
+	/**
+     * Create a string on memory.
+     * @param newLenght: lenght of the string to create.
+     * @param newString: C string.
+     */
+	void createString(const size_t& newLenght, const char* newString);
+
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: Pass `newLength` by value. In general, pass basic types by value

File: src/containers/inmutablestring.cpp [hunk 30]
```diff
@@ -1,5 +1,143 @@
...
+//////////////
+
+cmm::containers::InmutableString::~InmutableString()
+{
+	if (_string != nullptr)
+	{
+		cmm::utils::DebugUtilities::log(std::string("--- ~InmutableString()."));
+		stringLog();
+
+		delete(_string);
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: `delete _string` (See comments in `createString()`

File: src/utils/CMakeLists.txt [hunk 31]
```diff
@@ -0,0 +1,10 @@
+file (GLOB siminusminus_utils_sources
+		"*.cpp")
```
Comment (Manu343726) [inline src/utils/CMakeLists.txt]: **NERVER NERVER NEVER NERVER GLOB CMAKE SOURCES**. Explicitly list the sources in the target

**Review by Manu343726 — CHANGES_REQUESTED** [review]
The concept of the string class and its buffer management is mostly ok, but you have many little details to check. Please read the comments carefully. In essence:

 - **Do not commit commented code**
 - **Differentiate between local and global include directives**
 - **Learn the different GTest matchers**
 - **Learn the differences between a new/delete operator and a new/delete expression**

*There's also the tabs vs spaces thing, I'm an spaces guy (4 spaces indent actually), but this is a matter of taste...*

Also, in the future consider committing changes as you work, not to commit one week latter with thousands of changes. This makes the history more meaningful (Basically the resulting history resembles your way of thinking, the steps you've done to solve a task, the problems you encountered, etc) and makes things easier for the reviewer ;)

File: include/siminusminus/containers/inmutablestring.hpp (hunk 20 above)
Comment (ZaruWright) [reply include/siminusminus/containers/inmutablestring.hpp:50]: inmutablestring, I abbreviated the first word.

File: CMakeLists.txt (hunk 4 above)
Comment (ZaruWright) [reply CMakeLists.txt]: Only for curiosity. Why the flag name is NDEBUG and no DEBUG? is more efficient set macros on release than debug?

File: include/siminusminus/containers/inmutablestring.hpp (hunk 28 above)
Comment (ZaruWright) [reply include/siminusminus/containers/inmutablestring.hpp:30]: I don't know what you refer, there are some information about inmutable string just above of that line (You wrote that when you give me the hpp file). ¿You want that I rewrite something?

File: src/utils/debugutilities.cpp (hunk 9 above)
Comment (ZaruWright) [reply src/utils/debugutilities.cpp]: then, there isn't any difference about '/n' and std::endl?. I think its better std::endl because not depends on the platform you use.

File: src/utils/CMakeLists.txt (hunk 31 above)
Comment (ZaruWright) [reply src/utils/CMakeLists.txt]: Why?

File: include/siminusminus/containers/inmutablestring.hpp (hunk 22 above)
Comment (ZaruWright) [reply include/siminusminus/containers/inmutablestring.hpp:117]: I don't know what you refer

File: .travis.yml [hunk 32]
```diff
@@ -75,8 +75,8 @@ install:
   - if [ ! -d build ]; then mkdir build; fi
   - cd build
 
-  - cmake ..
-  - cmake --build .
+  - cmake .. -DBUILD_DOCUMENTATION=OFF
+  - cmake --build . -DBUILD_DOCUMENTATION=OFF
```
Comment (Manu343726) [inline .travis.yml]: Remove change in line 79, doesn't make sense

File: .travis.yml [hunk 33]
```diff
@@ -75,8 +75,8 @@ install:
   - if [ ! -d build ]; then mkdir build; fi
   - cd build
 
-  - cmake ..
-  - cmake --build .
+  - cmake .. -DBUILD_DOCUMENTATION=OFF
```
Comment (Manu343726) [inline .travis.yml]: Where's the code that implements this flag? Because it seems that the flag is being ignored. Are you sure it has been pushed?

**Review by Manu343726 — CHANGES_REQUESTED** [review]

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 34]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    InmutableIterator<T>& operator++()
+    {
+    	++_pointer;
+		return *this;
+    }
+
+    /**
+     * Post-increment operator that allows go over the object.
+     * @param rhs: Syntax operator to can different between pre and postincrement.
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: what? this sentence doesn't make sense to me

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 35]
```diff
@@ -0,0 +1,167 @@
...
+		
+	}
+
+	/**
+	 * Copy constructor.
+	 * @param iit: InmutableIterator<T> object that we'll copy from.
+	 */
+	InmutableIterator(const InmutableIterator<T>& iit): InmutableIterator<T>(*iit._pointer) 
+	{
+
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: remove this line

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 36]
```diff
@@ -0,0 +1,167 @@
...
+	/**
+	 * Default constructor. Init the iterator to null.
+	 */
+	InmutableIterator(): _pointer(nullptr)
+	{
+
+	}
+
+	/**
+	 * Constructor that takes a T object to point to object memory direction.
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: please check the grammar here

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 37]
```diff
@@ -0,0 +1,167 @@
...
+#include <algorithm>
+
+namespace cmm {
+namespace containers {
+
+/**
+ * \ingroup containers
+ * \brief iterator to inmutable objects.
+ * 
+ * The InmutableIterator class allows to create an iterator which cannot be
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: template, not class

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 38]
```diff
@@ -0,0 +1,167 @@
...
+ * modified, only allows read operations.
+ */
+template <typename T>
+class InmutableIterator : public std::iterator<std::input_iterator_tag, // iterator_category
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: The convention is to name const_iterator iterators with no write "rights", and iterator to iterator classes with write "rights". Since this template is not specific to a container but a generic iterator for contiguous data structures, I would name it ConstContiguousIterator or something like that.

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 39]
```diff
@@ -0,0 +1,167 @@
...
+	InmutableIterator(): _pointer(nullptr)
+	{
+
+	}
+
+	/**
+	 * Constructor that takes a T object to point to object memory direction.
+	 * @param object: object of T Type to init the iterator.
+	 */
+	InmutableIterator(const T& object) : _pointer(&object)
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: I'm not sure this is the expected behavior. I would make passing the address to the sequence element explicit

File: include/siminusminus/containers/inmutableiterator.hpp [hunk 40]
```diff
@@ -0,0 +1,167 @@
...
+	{
+		
+	}
+
+	/**
+	 * Copy constructor.
+	 * @param iit: InmutableIterator<T> object that we'll copy from.
+	 */
+	InmutableIterator(const InmutableIterator<T>& iit): InmutableIterator<T>(*iit._pointer)
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableiterator.hpp]: Following with the comment above, this line triggers all my C++ pedant alarms: Dereferencing the iterator, that is, reading the element pointed to, should only happen if the user requests to, because an iterator is completely agnostic about what it is currently pointing to. If you do this, you might introduce some potential bugs (Specifically, this copy constructor segfaults in the following case:

``` cpp
InmutableIterator<int> it;
InmutableIterator<int> it2 = it; // Boom!
```

which is really annoying since the two lines above are perfectly fine).

File: include/siminusminus/containers/inmutablestring.hpp [hunk 41]
```diff
@@ -59,14 +64,15 @@ class InmutableString
      * Returns the i th position from our string.
      * @param index: index to string access.
      */
-    char operator[](const size_t index) const;
+    char& operator[](const size_t index) const;
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: No, this change is wrong (And should not be compiling at all). It was ok in the previous version

File: include/siminusminus/containers/inmutablestring.hpp [hunk 42]
```diff
@@ -59,14 +64,15 @@ class InmutableString
...
      * An InmutableString is equal to another if their strings values
      * are the same.
      *
      * To compare and do other operations like ( >, <, >= and <=) we use the
-     * lexicographical order.
+     * lexicographical order. == is an exception doing with std::strcmp for
+     * efficiency.
      * @param rhs: the right hand side of the operation.
      */
     bool operator==(const InmutableString& rhs) const;
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: also add `operator!=`

File: include/siminusminus/containers/inmutablestring.hpp [hunk 43]
```diff
@@ -145,91 +163,12 @@ class InmutableString
...
-     * @param istring: InmutableString object.
-     */
-    InmutableStringIterator(const InmutableString& istring);
-
-    /**
-     * Copy constructor InmutableStringIterator from a same type object.
-     * @param istringit: InmutableStringIterator object.
-     */
-    InmutableStringIterator(const InmutableStringIterator& istringit);
+    InmutableIterator<char> _begin; // Iterator to the begining of the InmutableString
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutablestring.hpp]: remove, not needed. Create the iterators in `begin()` and ` end()` bodies

File: include/siminusminus/containers/inmutableviewer.hpp [hunk 44]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * \ingroup containers
+ * \brief Defines a range to iterate.
+ * 
+ * The InmutableViewer allows to iterate from a range of values. For example
+ * we can iterate between integers numbers as [5,10] or memory diretions like 
+ * a string also.
+ */
+template <typename T>
+class InmutableViewer : public InmutableIterator<T>
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableviewer.hpp]: View, not Viewer. Also check the comment in InmutableIterator

File: include/siminusminus/containers/inmutableviewer.hpp [hunk 45]
```diff
@@ -0,0 +1,68 @@
...
+{
+
+public:
+	/**
+	 * Constructor that takes two T objects, from and to, memory directions to iterate.
+	 * @param from: T object to start the range.
+	 * @param to: T object that ends the range.
+	 */
+	InmutableViewer(const T& from, const T& to): _begin(from), _end(to)
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableviewer.hpp]: Check the InmutableIterator comment again.

File: include/siminusminus/containers/inmutableviewer.hpp [hunk 46]
```diff
@@ -0,0 +1,68 @@
...
+	/**
+	 * Returns an InmutableIterator that points to the begining of the range.
+	 */
+	InmutableIterator<T> begin() const
+	{
+		return _begin;
+	}
+
+	/**
+	 * Returns an InmutableIterator that points to the ending of the range.
```
Comment (Manu343726) [inline include/siminusminus/containers/inmutableviewer.hpp]: to the end of

File: src/containers/inmutablestring.cpp [hunk 47]
```diff
@@ -2,77 +2,80 @@
...
+		delete[] _string;
 		setDefault(); // To take precautions
 	}
 	else
-		cmm::utils::DebugUtilities::log(std::string("--- ~InmutableString(). Nothing to delete, void string."));
+		DebugUtilities::log(std::string("--- ~InmutableString(). Nothing to delete, void string."));
 }
 
-char InmutableString::operator[](const size_t index) const
+char& InmutableString::operator[](const size_t index) const
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: undo this change

File: src/containers/inmutablestring.cpp [hunk 48]
```diff
@@ -2,77 +2,80 @@
...
+	return !std::lexicographical_compare(thisViewer.begin(), thisViewer.end(),
+										rhsViewer.begin(), rhsViewer.end());
 }
 
 bool InmutableString::operator<(const InmutableString& rhs) const
 {
-	InmutableStringIterator lhsit(this->_string);
-	InmutableStringIterator rhsit(rhs._string);
+	InmutableViewer<char> thisViewer(_begin, _end);
+	InmutableViewer<char> rhsViewer(rhs._begin, rhs._end);
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: why instancing views? just call the algorithm with your iterators

File: src/containers/inmutablestring.cpp [hunk 49]
```diff
@@ -87,13 +90,13 @@ bool InmutableString::operator<=(const InmutableString& rhs) const
...
-	cmm::utils::DebugUtilities::log(std::string("*** InmutableString::operator=(const InmutableString& rhs)."));
+	DebugUtilities::log(std::string("*** InmutableString::operator=(const InmutableString& rhs)."));
 
 	if (this == &rhs) // Points to the same data
 		return *this; // Not doing the assignment
 
-	if (this->_string) 
-		delete this->_string; // Erase data
+	if (_string) 
+		delete _string; // Erase data
```
Comment (Manu343726) [inline src/containers/inmutablestring.cpp]: delete[]

**Review by Manu343726 — CHANGES_REQUESTED** [review]
In addition to the comments, please check your indentation config. I would not approve code with random indents

File: src/containers/inmutablestring.cpp (hunk 49 above)
Comment (ZaruWright) [reply src/containers/inmutablestring.cpp]: _I believe you see another changes here, in the last version delete[] it's on._

Solved, i'm a little blind xdxdxd

File: include/siminusminus/containers/inmutablestring.hpp (hunk 41 above)
Comment (ZaruWright) [reply include/siminusminus/containers/inmutablestring.hpp]: I have been reading this [ std::string operator[] ](http://es.cppreference.com/w/cpp/string/basic_string/operator_at) and returns a reference. I suppose it's wrong because you can modify the character and an InmutableString don't allow that, but if I return a const reference? I need it to do something like this: 
`
InmutableString istring("Hello");
InmutableIterator<char> iitbegin (&istring[0]);`

File: src/containers/inmutablestring.cpp (hunk 47 above)
Comment (ZaruWright) [reply src/containers/inmutablestring.cpp]: Waiting above answer
